drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,945 @@
+		tmp = (tmp & PDS_CLKPLL_XTAL_RC32M_SEL_UMSK) |
+		      (0U << PDS_CLKPLL_XTAL_RC32M_SEL_POS);
+	} else {
+		tmp = (tmp & PDS_CLKPLL_REFCLK_SEL_UMSK) | (0U << PDS_CLKPLL_REFCLK_SEL_POS);
+		tmp = (tmp & PDS_CLKPLL_XTAL_RC32M_SEL_UMSK) |
+		      (1U << PDS_CLKPLL_XTAL_RC32M_SEL_POS);
+	}
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+}
+
+static void clock_control_bl60x_init_pll(enum bl60x_clkid source, uint32_t crystal_frequency)
+{
+	uint32_t tmp;
+	uint32_t old_rootclk;
+
+	old_rootclk = clock_control_bl60x_get_root_clock();
+
+	/* security RC32M */
+	if (old_rootclk > 1) {
+		clock_control_bl60x_set_root_clock(0);
+	}
+
+	clock_control_bl60x_deinit_pll();
+
+	if (source == BL60X_CLKID_CLK_CRYSTAL) {
+		clock_control_bl60x_set_pll_source(1);
+	} else {
+		clock_control_bl60x_set_pll_source(0);
+	}
+
+	/* 26M special treatment */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_CP_OFFSET);
+	if (clock_control_bl60x_crystal_to_id(crystal_frequency) == 4) {
+		tmp = (tmp & PDS_CLKPLL_ICP_1U_UMSK) | (1U << PDS_CLKPLL_ICP_1U_POS);
+		tmp = (tmp & PDS_CLKPLL_ICP_5U_UMSK) | (0U << PDS_CLKPLL_ICP_5U_POS);
+		tmp = (tmp & PDS_CLKPLL_INT_FRAC_SW_UMSK) | (1U << PDS_CLKPLL_INT_FRAC_SW_POS);
+	} else {
+		tmp = (tmp & PDS_CLKPLL_ICP_1U_UMSK) | (0U << PDS_CLKPLL_ICP_1U_POS);
+		tmp = (tmp & PDS_CLKPLL_ICP_5U_UMSK) | (2U << PDS_CLKPLL_ICP_5U_POS);
+		tmp = (tmp & PDS_CLKPLL_INT_FRAC_SW_UMSK) | (0U << PDS_CLKPLL_INT_FRAC_SW_POS);
+	}
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_CP_OFFSET);
+
+	/* More 26M special treatment */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_RZ_OFFSET);
+	if (clock_control_bl60x_crystal_to_id(crystal_frequency) == 4) {
+		tmp = (tmp & PDS_CLKPLL_C3_UMSK) | (2U << PDS_CLKPLL_C3_POS);
+		tmp = (tmp & PDS_CLKPLL_CZ_UMSK) | (2U << PDS_CLKPLL_CZ_POS);
+		tmp = (tmp & PDS_CLKPLL_RZ_UMSK) | (5U << PDS_CLKPLL_RZ_POS);
+		tmp = (tmp & PDS_CLKPLL_R4_SHORT_UMSK) | (0U << PDS_CLKPLL_R4_SHORT_POS);
+	} else {
+		tmp = (tmp & PDS_CLKPLL_C3_UMSK) | (3U << PDS_CLKPLL_C3_POS);
+		tmp = (tmp & PDS_CLKPLL_CZ_UMSK) | (1U << PDS_CLKPLL_CZ_POS);
+		tmp = (tmp & PDS_CLKPLL_RZ_UMSK) | (1U << PDS_CLKPLL_RZ_POS);
+		tmp = (tmp & PDS_CLKPLL_R4_SHORT_UMSK) | (1U << PDS_CLKPLL_R4_SHORT_POS);
+	}
+	tmp = (tmp & PDS_CLKPLL_R4_UMSK) | (2U << PDS_CLKPLL_R4_POS);
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_RZ_OFFSET);
+
+	/* set pll dividers */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_POSTDIV_UMSK) | ((uint32_t)(0x14) << PDS_CLKPLL_POSTDIV_POS);
+	tmp = (tmp & PDS_CLKPLL_REFDIV_RATIO_UMSK) | (2U << PDS_CLKPLL_REFDIV_RATIO_POS);
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+
+	/* set SDMIN */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_SDM_OFFSET);
+	if (source == BL60X_CLKID_CLK_CRYSTAL) {
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK) |
+		      (clock_control_bl60x_crystal_SDMIN_table[clock_control_bl60x_crystal_to_id(
+			       crystal_frequency)]
+		       << PDS_CLKPLL_SDMIN_POS);
```

Comment:
`clock_control_bl60x_crystal_to_id()` may return `-1` which causes undefined behavior when passed to an array, why dont we just map the freq to id as we read from the DT?

```c
#define CRYSTAL_ID_FREQ_32000000 0
#define CRYSTAL_ID_FREQ_24000000 1
#define CRYSTAL_ID_FREQ_38400000 2
#define CRYSTAL_ID_FREQ_40000000 3
#define CRYSTAL_ID_FREQ_26000000 4

#define CRYSTAL_FREQ_TO_ID(freq) CONCAT(CRYSTAL_ID_FREQ_, freq)

CRYSTAL_FREQ_TO_ID(DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, crystal), clock_frequency))
```

If the configuration is wrong, it won't compile

Reply:
You mean by using an enum inside devicetree binding, right?
Then user can only select those values otherwise devicetree will already report an error.

Reply:
it wont compile because the define doesnt exist for frequency X

Reply:
done